File: Java/lesson 15/CountDistinctSlices.java
// you can also use imports, for example:
import java.util.*;

// you can write to stdout for debugging purposes, e.g.
// System.out.println("this is a debug message");

class Solution {
    public int solution(int M, int[] A) {
        // write your code in Java SE 8
        int res = 0;
        Set<Integer> set = new HashSet<Integer>();
        for (int i = 0, j = 0; i < A.length; i++) {
            while (j < A.length && !set.contains(A[j])) {
                set.add(A[j]);
                j++;
            }
            res += j - i;
            if (res > 1000000000) return 1000000000;
            set.remove(A[i]);
        }
        return res;
    }
}

File: Go/lesson 15/CountDistinctSlices.go
package solution

// you can also use imports, for example:
// import "fmt"
// import "os"

// you can write to stdout for debugging purposes, e.g.
// fmt.Println("this is a debug message")

func Solution(M int, A []int) int {
    // write your code in Go 1.4
    res := 0
    set := make([]int, M + 1)
    for i, j := 0, 0; i < len(A); i++ {
        for j < len(A) && set[A[j]] == 0 {
            set[A[j]] = 1
            j++
        }
        res += j - i
        if res > 1000000000 {return 1000000000}
        set[A[i]] = 0
    }
    return res
}